src/test-explorer/add-test-suite.ts

```diff
@@ -0,0 +1,104 @@
+import * as vscode from "vscode";
+import { testCache } from "./test-cache";
+import { TargetName, TargetUri, TestItemMetadata, AddTestSuite } from "./types";
+import { prefixesOf, toVscodeRange } from "./util";
+
+/**
+ * Add test suite to the Test Explorer.
+ *
+ * Depending on fully qualified name traverse Test Explorer hierarchy,
+ * creating intermediate test items if necessary, and find test item for which
+ * a newly created test suite item should be added.
+ *
+ * If test item already exists do nothing
+ */
+export function addTestSuite(
+  testController: vscode.TestController,
+  targetName: TargetName,
+  targetUri: TargetUri,
+  event: AddTestSuite
+): void {
+  const buildTargetItem = getOrCreateBuildTargetItem(
+    testController,
+    targetName,
+    targetUri
+  );
+
+  function addTestSuiteLoop(parent: vscode.TestItem, testIds: string[]) {
+    if (testIds.length > 0) {
+      const [currentId, ...restOfIds] = testIds;
+      const child = parent.children.get(currentId);
+      if (child) {
+        addTestSuiteLoop(child, restOfIds);
+      } else {
+        const packageNode = testController.createTestItem(currentId, currentId);
+        parent.children.add(packageNode);
+
+        const data: TestItemMetadata = {
+          kind: "package",
+          targetName,
+          targetUri,
+        };
+        testCache.setMetadata(packageNode, data);
+        addTestSuiteLoop(packageNode, restOfIds);
+      }
+    } else {
+      const { className, location, fullyQualifiedClassName } = event;
+
+      if (parent.children.get(fullyQualifiedClassName)) {
```

Comment:
Shouldn't we replace it if the suite already exists?

Reply:
I think we should